src/shared/sol-conffile.c

```diff
@@ -152,53 +260,62 @@ sol_conffile_load_keyfile_from_dirs(const char *file, char **full_path)
     else
         SOL_DBG("could not getcwd: %d", errno);
 
-    keyfile = g_key_file_new();
-    SOL_NULL_CHECK(keyfile, NULL);
-
     /* absolute path */
-    if (g_path_is_absolute(file)) {
-        if (!g_key_file_load_from_file(keyfile,
-            file, G_KEY_FILE_NONE, &error))
-            goto err;
-    } else if (!g_key_file_load_from_dirs(keyfile, file, search_dirs,
-        full_path, G_KEY_FILE_NONE,
-        &error))
-        goto err;
-
-    return keyfile;
-
-err:
-    SOL_DBG("could not load file '%s' from dirs: %s",
-        file, error ? error->message : "unknown error");
-    g_error_free(error);
-    g_key_file_free(keyfile);
-    return NULL;
+    if (file[0] == '/') {
+        keyfile = fopen(file, "r");
+    } else {
+        for (i = 0; i < search_dirs_size; i++) {
+            char filename[PATH_MAX+1];
+            struct stat st;
+            sprintf(filename, "%s/%s", search_dirs[i],file);
+            if (stat(filename, &st) != 0)
+                continue;
+
+            keyfile = fopen(filename, "r");
+            if(keyfile) {
+                break;
+            }
+        }
+    }
+
+    if (!keyfile)
+        SOL_DBG("could not load file '%s' from dirs: unknown error");
+
+    /* reads all the config to a temporary buffer */
+    fseek (keyfile, 0, SEEK_END);
+    length = ftell (keyfile);
+    fseek (keyfile, 0, SEEK_SET);
+    config_file_contents = malloc (length);
+    if (config_file_contents) {
+        fread (config_file_contents, 1, length, keyfile);
+    }
+    fclose (keyfile);
+    return config_file_contents;
 }
 
-static GKeyFile *
+static char*
 sol_conffile_load_keyfile_from_paths(const char *path,
-    const char *fallback_paths,
-    char **full_path)
+            const char *fallback_paths,
+            char **full_path)
 {
-    gchar **ptr;
-    gchar **splitted_paths;
-    GKeyFile *keyfile = NULL;
+    char **ptr;
+    char **splitted_paths;
+    char *config_file_contents = NULL;
 
     if (path)
-        keyfile = sol_conffile_load_keyfile_from_dirs(path, full_path);
+        config_file_contents = sol_conffile_load_keyfile_from_dirs(path, full_path);
 
-    if (fallback_paths && !keyfile) {
-        splitted_paths = g_strsplit(fallback_paths, ";", -1);
+    if (fallback_paths && !config_file_contents) {
+        splitted_paths = str_split(fallback_paths, ';');
         for (ptr = splitted_paths; *ptr; ptr++) {
-            keyfile = sol_conffile_load_keyfile_from_dirs
-                    (*ptr, full_path);
-            if (keyfile)
+            config_file_contents = sol_conffile_load_keyfile_from_dirs(*ptr, full_path);
+            if (config_file_contents) {
                 break;
+            }
         }
-        g_strfreev(splitted_paths);
+        free (splitted_paths);
```

Comment:
`g_strfreev()` is not equivalent to `free()`. It will free the memory block pointed to `splitted_paths`, but before that, it'll also free each element it points to. (It stands for `free vector`.)